src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs

```diff
@@ -0,0 +1,204 @@
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.ResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static IDisposable? BeginRequestPipelineScope(this ILogger logger, HttpRequestMessage request)
+        {
+            return _beginRequestPipelineScope(logger, request.Method, GetRedactedUriString(request.RequestUri));
+        }
+
+        public static void LogRequestPipelineStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineRequestHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Request, request.Headers, request.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static void LogRequestPipelineEnd(this ILogger logger, HttpResponseMessage response, TimeSpan duration, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineEnd(logger, duration.TotalMilliseconds, (int)response.StatusCode, null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestPipelineResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        internal static string? GetRedactedUriString(Uri? uri)
+        {
+            if (uri is null)
+            {
+                return null;
+            }
+
+            if (!uri.IsAbsoluteUri)
+            {
+                // We cannot guarantee the redaction of UserInfo for relative Uris without implementing some subset of Uri parsing in this package.
+                // To avoid this, we redact the whole Uri. Seeing a relative Uri in LoggingHttpMessageHandler or LoggingScopeHttpMessageHandler
+                // requires a custom handler chain with custom expansion logic implemented by the user's HttpMessageHandler.
+                // In such advanced scenarios we recommend users to log the Uri in their handler.
+                return "*";
+            }
+
+            string uriString = uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.UserInfo, UriFormat.UriEscaped);
```

Comment:
I just remembered this won't actually work since Uris created with `DangerousDisablePathAndQueryCanonicalization` won't let you ask for the path/query via `GetComponents` (we might have to rethink that).

A low-alloc version would look something like this:
```c#
private static string GetRedactedUriString(Uri? uri)
{
    if (uri is null) return null;
    
    if (!uri.IsAbsoluteUri) return "*";

    string pathAndQuery = uri.PathAndQuery;
    int queryIndex = s_disableUriQueryRedaction ? -1 : pathAndQuery.IndexOf('?');

    return (queryIndex >= 0, uri.IsDefaultPort) switch
    {
        (true, true) => $"{uri.Scheme}://{uri.Host}{pathAndQuery.AsSpan(0, queryIndex + 1)}*",
        (true, false) => $"{uri.Scheme}://{uri.Host}:{uri.Port}{pathAndQuery.AsSpan(0, queryIndex + 1)}*",
        (false, true) => $"{uri.Scheme}://{uri.Host}{pathAndQuery}",
        (false, false) => $"{uri.Scheme}://{uri.Host}:{uri.Port}{pathAndQuery}"
    };
}
```

Note that this also removes the fragment. Do we care about it? It's unlikely to be set and won't be sent over the wire anyway.

Reply:
>Do we care about it? It's unlikely to be set and won't be sent over the wire anyway.

I would prefer to have consistency between the redaction here and redaction for `url.full` implemented in https://github.com/dotnet/runtime/pull/104251, where I wanted to be compatible with the [OTel logic](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/4c69afa2fbb3b9d84b45bd73dcfad43331516e69/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpTagHelper.cs#L21) which does emit the fragment. But given that our query redaction is already incompatible we can probably omit fragment as well. I will ping Vishwesh on this.